libraries/AP_Math/quaternion.cpp

```diff
@@ -118,6 +118,271 @@ void Quaternion::from_rotation_matrix(const Matrix3f &m)
+        q1 = q2 = 0.27059802f;
+        q3 = 0.65328145f;
+        q4 = 0.65328151f;
+        return;
+
+    case ROTATION_ROLL_270:
+        q1 = HALF_SQRT_2;
+        q2 = -HALF_SQRT_2;
+        q3 = q4 = 0;
+        return;
+
+    case ROTATION_ROLL_270_YAW_45:
+        q1 = 0.65328151f;
+        q2 = -0.65328145f;
+        q3 = -0.27059802f;
+        q4 = 0.27059802f;
+        return;
+
+    case ROTATION_ROLL_270_YAW_90:
+        q1 = q4 = 0.5f;
+        q2 = q3 = -0.5f;
+        return;
+
+    case ROTATION_ROLL_270_YAW_135:
+        q1 = 0.27059802f;
+        q2 = -0.27059802f;
+        q3 = -0.65328145f;
+        q4 = 0.65328151f;
+        return;
+
+    case ROTATION_PITCH_90:
+        q1 = q3 = HALF_SQRT_2;
+        q2 = q4 = 0;
+        return;
+
+    case ROTATION_PITCH_270:
+        q1 = HALF_SQRT_2;
+        q2 = q4 = 0;
+        q3 = -HALF_SQRT_2;
+        return;
+
+    case ROTATION_PITCH_180_YAW_90:
+        q1 = q4 = 0;
+        q2 = -HALF_SQRT_2;
+        q3 = HALF_SQRT_2;
+        return;
+
+    case ROTATION_PITCH_180_YAW_270:
+        q1 = q4 = 0;
+        q2 = q3 = HALF_SQRT_2;
+        return;
+
+    case ROTATION_ROLL_90_PITCH_90:
+        q1 = q2 = q3 = -0.5f;
+        q4 = 0.5f;
+        return;
+
+    case ROTATION_ROLL_180_PITCH_90:
+        q1 = q3 = 0;
+        q2 = -HALF_SQRT_2;
+        q4 = HALF_SQRT_2;
+        return;
+
+    case ROTATION_ROLL_270_PITCH_90:
+        q1 = q3 = q4 = 0.5f;
+        q2 = -0.5f;
+        return;
+
+    case ROTATION_ROLL_90_PITCH_180:
+        q1 = q2 = 0;
+        q3 = -HALF_SQRT_2;
+        q4 = HALF_SQRT_2;
+        return;
+
+    case ROTATION_ROLL_270_PITCH_180:
+        q1 = q2 = 0;
+        q3 = q4 = HALF_SQRT_2;
+        return;
+
+    case ROTATION_ROLL_90_PITCH_270:
+        q1 = q2 = q4 = 0.5f;
+        q3 = -0.5;
+        return;
+
+    case ROTATION_ROLL_180_PITCH_270:
+        q1 = q3 = 0;
+        q2 = q4 = HALF_SQRT_2;
+        return;
+
+    case ROTATION_ROLL_270_PITCH_270:
+        q1 = -0.5f;
+        q2 = q3 = q4 = 0.5f;
+        return;
+
+    case ROTATION_ROLL_90_PITCH_180_YAW_90:
+        q1 = q3 = -0.5f;
+        q2 = q4 = 0.5f;
+        return;
+
+    case ROTATION_ROLL_90_YAW_270:
+        q1 = q2 = -0.5f;
+        q3 = q4 = 0.5f;
+        return;
+
+    case ROTATION_ROLL_90_PITCH_68_YAW_293:
+        q1 = 0.26774535f;
+        q2 = 0.70698798f;
+        q3 = 0.01295743f;
+        q4 = -0.65445596f;
+        return;
+
+    case ROTATION_PITCH_315:
+        q1 = 0.92387956f;
+        q2 = q4 = 0;
+        q3 = -0.38268343f;
+        return;
+
+    case ROTATION_ROLL_90_PITCH_315:
+        q1 = 0.65328151f;
+        q2 = 0.65328145f;
+        q3 = -0.27059802f;
+        q4 = 0.27059802f;
+        return;
+
+    case ROTATION_PITCH_7:
+        q1 = 0.99813479f;
+        q2 = q4 = 0;
+        q3 = 0.06104854f;
+        return;
+
+    case ROTATION_MAX:
+    case ROTATION_CUSTOM:
+        // no-op; custom rotations not supported
```

Comment:
This looks like a good place to put an `AP_InternalError::error(AP_InternalError::error_t::flow_of_control`

Reply:
ok added, thanks